rfcs/0181-list-index-syntax.md

```diff
@@ -0,0 +1,290 @@
+* `expr.prefix.[n].suffix or expr2` is the result of evaluating `if expr ? prefix.[n].suffix then expr.prefix.[n].suffix else expr2`
+
+It is a syntax error if either new production is used in the left-hand side of a binding.
+
+An implementation of this design is available as patches for Nix at <https://gitlab.com/rhendric/nix-list-index-syntax/>; see instructions there for use.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+```
+nix-repl> [ 1 4 9 ].[2]
+9
+
+nix-repl> pkgs.maptool.meta.maintainers.[0].github
+"rhendric"
+
+nix-repl> [ 1 4 9 ].[3]
+error: list index 3 is out of bounds
+
+       at «string»:1:1:
+
+            1| [ 1 4 9 ].[3]
+             | ^
+
+nix-repl> [ 1 4 9 ] ? [2]
+true
+
+nix-repl> [ 1 4 9 ] ? [3]
+false
+
+nix-repl> [ 1 4 9 ] ? [0].[0]
+false
+
+nix-repl> [ [ 1 4 9 ] ] ? [0].[0]
+true
+```
+
+The following are all syntax errors:
+
+```
+nix-repl> let n = 0; in [ 4 ].[n]
+error: syntax error, unexpected ID, expecting INT
+
+       at «string»:1:22:
+
+            1| let n = 0; in [ 4 ].[n
+             |                      ^
+
+nix-repl> { a.[0] = true; }
+error: syntax error, index '0' not allowed here
+
+       at «string»:1:3:
+
+            1| { a.[0] = true;
+             |   ^
+
+nix-repl> let a.[0] = true; in a                   
+error: syntax error, index '0' not allowed here
+
+       at «string»:1:5:
+
+            1| let a.[0] = true;
+             |     ^
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The prototype [patch] implementing this feature adds a net 69 lines of code to Nix, excluding tests.
+
+[patch]: https://gitlab.com/rhendric/nix-list-index-syntax/-/blob/main-dist/patches/bracketed/from-90e630a5.patch
+
+As noted in passing in the motivation section, the proposed syntax differs from the syntax already used by Nix tools on the command line for `--attr`.
+`nix-instantiate` has no difficulty with:
+
+```
+$ nix-instantiate --eval '<nixpkgs>' -A maptool.meta.maintainers.0.github
+"rhendric"
+```
+
+This divergence might lead to confusion.
+
+Evolving the syntax of Nix always imposes a cost on third-party tools that process Nix syntax, including syntax highlighters, linters, formatters, static analyzers, and language servers.
+This syntax is not a dramatic extension of the language but would require support from all of the above for them to maintain full functionality.
+
+Finally, there is an opportunity cost to claiming new syntax.
+One could imagine speculative features that might want to use this syntax, such as a list or string slicing syntax, or a ‘list swizzle’ operator that desugars `expr.[ 2 0 1 ]` to `[ (elemAt expr 2) (elemAt expr 0) (elemAt expr 1) ]`.
+It is, in my opinion, unlikely that list and string manipulation (assuming that any feature in competition for this syntax would involve lists or strings somehow) would be so common in Nix to make this a compelling objection.
+
+# Alternatives
```

Comment:
Another alternative which might be worth mentioning would be `foo |> getElem 2` using the pipe operator (RFC #148). I think this may work sufficiently okay in a REPL, however in Nix code this may interfere with existing `|>` pipelines and thus require parentheses

Reply:
Even in a REPL, the motivating case of `(foo.bar.qux |> getElem 0).moreStuff` requires parentheses.

Reply:
> Another alternative which might be worth mentioning would be `foo |> getElem 2` using the pipe operator

I agree that "pipe operator" is good but not a real alternative of this case.